rfcs/0006-gtfs-canonical-routes.md

```diff
@@ -0,0 +1,125 @@
+  case for bus initially).
+- **shapes.txt** will add new shapes for all canonical subway trips, since the shape IDs can change
+  between ratings, especially when stations are open and closed on a long-term basis. These new
+  shapes can have an identical set of points as those shapes used during current service. Existing
+  Commuter Rail shape references can be reused for canonical trips.
+- **trips.txt** includes one new trip for each canonical route/direction pair, or multiple trips in
+  the event of regular branching occuring along the route. It should reference the new `service_id`
+  created above. It should reference either a new (canonical) or previously-existing (regular
+  service) route pattern, depending on whether the trip's stopping pattern is currently served by
+  regular service. Each of these trips should be the `representative_trip_id` for their referenced
+  route patterns.
+- **stop_times.txt** includes all stops that trains in the route/direction passes through, even if
+  they are temporarily closed. Increment the arrival/departure times by one minute for each
+  subsequent stop. For a sample, see the attached GTFS file, the output of [✓ 🧪 🚝 Canonical
+  trips/patterns experiment](https://app.asana.com/0/881264583703207/1200210504369250), which
+  trialled this for Green Line D and Lowell Line (only in the outbound direction).
+
+This will be done as a new "step" of the Makefile called `canonical`, which runs between `truncate`
+and `current`. There will be a new a directory, `input/canonical/` to store any static data the step
+uses.
+
+The canonical step will be a Python module that reads in the `truncate` feed, and writes out its own
+feed after transforming it. The transformations are as above, and also will update any canonical
+trips to use current route patterns (and vice versa) if relevant, eliminating redundant route
+patterns.
+
+For Commuter Rail route patterns, the canonical one will specify every stop as non-flag, non-leave
+early stop. For both subway and Commuter Rail canonical trips, no data around bicycle allowances
+will be provided (`bikes_allowed=0`).
+
+# Drawbacks
+
+This is more complicated than not doing it.
+
+# Rationale and alternatives
+
+See proposed Concepts [1](https://github.com/mbta/technology-docs/pull/6#issuecomment-952315958), [4](https://github.com/mbta/technology-docs/pull/6#issuecomment-952315958), and [5](https://github.com/mbta/technology-docs/pull/6#issuecomment-962120434) in prior discussion.
+
+# Prior art
+
+Until now we've been doing ad-hoc adjustments to determine the route pattern of routes
+[here](https://github.com/mbta/api/blob/master/apps/state/config/config.exs#L145).
+
+# Unresolved questions
+
+Ensuring that we would not be in a situation where the sort order, perhaps ordered by
+frequency/proportion of trips, becomes a different order than desired for the main route pattern
+against the branch route pattern.
```

Comment:
Is this sort-of resolved now that we've specified the changes to `route_pattern_sort_order`? It's missing an explicit statement that if existing patterns are deemed canonical their relative sort order should be updated to match the intended one, but this seems feasible in principle.

Reply:
Like with the next comment below, I wanted to check whether anyone had any qualms with the `route_pattern_sort_order` adjustment I proposed.

>It's missing an explicit statement that if existing patterns are deemed canonical their relative sort order should be updated to match the intended one, but this seems feasible in principle.

_Should_ such existing patterns have their sort order updated in some way? The only goal I had accounted for in the latest text was ensuring that canonical patterns across different branches would be sorted correctly, regardless of whether "a new fake pattern" or "an existing pattern" is utilized.

Reply:
> The only goal I had accounted for in the latest text was ensuring that canonical patterns across different branches would be sorted correctly, regardless of whether "a new fake pattern" or "an existing pattern" is utilized.

This suggests to me the answer to your question is "yes", because there's no guarantee (is there?) that existing patterns which are marked canonical will have the same relative sort orders as the artificial canonical ones we generate. But it's very possible I'm misunderstanding you, and I think there's some vagueness in the current RFC wording:

> This will be done by adding two new digits to `route_pattern_sort_order` (one for branch and the other for canonicalness, both placed right after the existing digits corresponding to the `route_sort_order`)

Could you give a concrete example of this approach? How exactly is the digit for "branch" determined, both for our "constructed" canonical patterns and the real ones? Is the digit for "canonicity" set for constructed patterns and not set for real ones that are marked canonical, or is it set for both?

Reply:
<img width="857" alt="Screen Shot 2021-12-21 at 13 42 52" src="https://user-images.githubusercontent.com/3793006/147001310-e41c6f1a-4bd6-47e6-9abb-d3a00509e235.png">

Above is an example I had put together when thinking through the proposed solution. (The illegal non-digits in the `route_pattern_sort_order`s I put in for temporary clarity.)

The second block (right after the "blah", which represents the corresponding `route_sort_order`) would be set in GTFS-creator based on the branch ordering, which itself would be hardcoded. For instance, for the Red Line, we could say that any pattern that touches "Savin Hill" or "Ashmont" is on the `3` branch, and any pattern touching "North Quincy" or "Braintree" is on the `4` branch.

> Is the digit for "canonicity" set for constructed patterns and not set for real ones that are marked canonical[?]

Yes on this part of the question, or at least they would be set differently, so that any exposed "constructed patterns" get sorted to the end of a branch's patterns.

⚠️ The one case where this may look a little funny is if a client is sorting route patterns exclusively on `route_pattern_sort_order`, without regard for `route_pattern_typicality` or `canonical_route_pattern`. In that case, canonical "constructed" patterns would be mixed in among in-service patterns for multi-branch routes. This is a small subset of routes, though I'd suggest mitigating this by recommending, in the [GTFS-documentation](https://github.com/mbta/gtfs-documentation/blob/master/reference/gtfs.md#route_patternstxt), to use `route_pattern_sort_order` in conjunction with the other fields, much as dotcom's existing practice.

Reply:
Is there a reason that we should require consumers to consider other fields besides `route_pattern_sort_order` for sorting? People don't read documentation, so it would be more convenient for them (and easier for us to update in the future) if they only need to consider one field.

Reply:
I'd also argue for `sort_order` to be all-inclusive. It's easier to understand that way, and if we already know how we want clients to sort patterns, we should do it for them — that way the algorithm stays with us, instead of being encoded across many different clients where it's much harder to change.

Reply:
I might push back somewhat on the idea that we know how _all_ clients want to display patterns. Some clients/applications may want to show unscheduled canonical patterns, and others might not. This could (and does) occur even with our current setup, if a client wanted to obscure highly-atypical patterns into an "Uncommon destinations". But I think @paulswartz's comment stands in that, given that we don't know exactly how a given client wants to show a route's patterns, we should provide these fields as-is without recommendation as to _how_ to combine them...just as we do now.

Reply:
@digitalcora I think we can add something to GTFS-creator's truncation phase such that it checks, for these select branched routes, for these these conditions, given that a **Branch A** should be "shown" first before a **Branch B**, automatically adjusting sort orders as needed to meet them:

- Among all `canonical_route_pattern=1` patterns on the route/direction, all those on **Branch A** must have `route_pattern_sort_order`s less in value than all those on **Branch B**.
- Among all `route_pattern_typicality<5` patterns on the route/direction, at least one pattern on **Branch A** must have a `route_pattern_sort_order` less in value than the **Branch B** pattern with the smallest `route_pattern_sort_order`.
- No patterns should have the same `route_pattern_sort_order` between two different `direction_id`s on the same `route_id`.

Do these seem reasonable? If that's the case, I'll remove explicit mention of adding new digits to `route_pattern_sort_order` (if that's truly needed, especially to handle the third constraint, it can be added as part of the implementation).

Reply:
Sorry for the basic question, but can someone give me an example where `route_pattern_sort_order` is used? I'm having trouble reifying this discussion. I understand the sort order of canonical patterns for determining the "main branch", but can't think of other places I've seen all the patterns in some sorted order.

Reply:
The website displays patterns in an order, but I don't know whether they use `route_pattern_sort_order` or not: https://www.mbta.com/schedules/9/line?schedule_direction%5Bdirection_id%5D=0&schedule_direction%5Bvariant%5D=9-_-0 (click the "City Point" dropdown to see them)

Reply:
They're sorted [by typicality and then by sort order](https://github.com/mbta/dotcom/blob/74a4703/apps/route_patterns/lib/repo.ex#L44).